plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchange.java

```diff
@@ -199,53 +208,39 @@ private List<ExchangeSourceHandle> createExchangeSourceHandles()
-            ImmutableMultimap.Builder<Integer, FileStatus> result = ImmutableMultimap.builder();
-            for (FileStatus partitionFile : partitionFiles) {
-                Matcher matcher = PARTITION_FILE_NAME_PATTERN.matcher(new File(partitionFile.getFilePath()).getName());
-                checkState(matcher.matches(), "Unexpected partition file: %s", partitionFile);
-                int partitionId = Integer.parseInt(matcher.group(1));
-                result.put(partitionId, partitionFile);
-            }
-            return result.build();
-        }
-        catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
+        URI sinkOutputPath = getTaskOutputDirectory(taskPartitionId);
+        return Futures.transform(
+                exchangeStorage.listFilesRecursively(sinkOutputPath),
+                sinkOutputFiles -> {
+                    String committedMarkerFilePath = sinkOutputFiles.stream()
+                            .map(FileStatus::getFilePath)
+                            .filter(filePath -> filePath.endsWith(COMMITTED_MARKER_FILE_NAME))
+                            .findFirst()
+                            .orElseThrow(() -> new IllegalStateException(format("No committed attempts found under sink output path %s", sinkOutputPath)));
+                    // Committed marker file path format: {sinkOutputPath}/{attemptId}/committed
+                    String[] parts = committedMarkerFilePath.split(PATH_SEPARATOR);
+                    checkState(parts.length >= 3, "committedMarkerFilePath %s is malformed", committedMarkerFilePath);
+                    String committedAttemptId = parts[parts.length - 2];
+                    int attemptIdOffset = committedMarkerFilePath.length() - committedAttemptId.length()
+                            - PATH_SEPARATOR.length() - COMMITTED_MARKER_FILE_NAME.length();
+
+                    // Data output file path format: {sinkOutputPath}/{attemptId}/{sourcePartitionId}_{splitId}.data
+                    List<FileStatus> partitionFiles = sinkOutputFiles.stream()
+                            .filter(file -> file.getFilePath().startsWith(committedAttemptId + PATH_SEPARATOR, attemptIdOffset) && file.getFilePath().endsWith(DATA_FILE_SUFFIX))
+                            .collect(toImmutableList());
+
+                    ImmutableMultimap.Builder<Integer, FileStatus> result = ImmutableMultimap.builder();
+                    for (FileStatus partitionFile : partitionFiles) {
+                        Matcher matcher = PARTITION_FILE_NAME_PATTERN.matcher(new File(partitionFile.getFilePath()).getName());
+                        checkState(matcher.matches(), "Unexpected partition file: %s", partitionFile);
+                        int partitionId = Integer.parseInt(matcher.group(1));
+                        result.put(partitionId, partitionFile);
+                    }
+                    return result.build();
+                },
+                directExecutor());
```

Comment:
I wonder if it makes sense to run it in the executor to avoid potentially blocking an important IO thread of an underlying library

Reply:
Why is it? The transform part only involves code execution, no external API calls, right?

Reply:
It does loops, regex matching, etc. Ideally it is better not to run anything even moderately resource consuming as it may steal important resources from rather (usually) tiny IO pool.

Reply:
OK, I'm changing it to use `executor`